eppoclient/ufc.go

```diff
@@ -0,0 +1,143 @@
+package eppoclient
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+type ufcResponse struct {
+	Flags map[string]flagConfiguration `json:"flags"`
+}
+
+type flagConfiguration struct {
+	Key           string               `json:"key"`
+	Enabled       bool                 `json:"enabled"`
+	VariationType variationType        `json:"variationType"`
+	Variations    map[string]variation `json:"variations"`
+	Allocations   []allocation         `json:"allocations"`
+	TotalShards   int64                `json:"totalShards"`
+}
+
+type variation struct {
+	Key   string      `json:"key"`
+	Value interface{} `json:"value"`
+}
+
+type variationType int
+
+const (
+	stringVariation variationType = iota
+	integerVariation
+	numericVariation
+	booleanVariation
+	jsonVariation
+)
+
+func (v variationType) MarshalJSON() ([]byte, error) {
+	switch v {
+	case stringVariation:
+		return json.Marshal("STRING")
+	case integerVariation:
+		return json.Marshal("INTEGER")
+	case numericVariation:
+		return json.Marshal("NUMERIC")
+	case booleanVariation:
+		return json.Marshal("BOOLEAN")
+	case jsonVariation:
+		return json.Marshal("JSON")
+	default:
+		return nil, fmt.Errorf("unsupported variation type: %d", v)
+	}
+}
+
+func (v *variationType) UnmarshalJSON(data []byte) error {
+	var value string
+	if err := json.Unmarshal(data, &value); err != nil {
+		return err
+	}
+
+	switch value {
+	case "STRING":
+		*v = stringVariation
+	case "INTEGER":
+		*v = integerVariation
+	case "NUMERIC":
+		*v = numericVariation
+	case "BOOLEAN":
+		*v = booleanVariation
+	case "JSON":
+		*v = jsonVariation
+	default:
+		return fmt.Errorf("unknown variation type: %s", value)
+	}
+
+	return nil
+}
+
+func (ty variationType) valueToAssignmentValue(value interface{}) (interface{}, error) {
+	switch ty {
+	case stringVariation:
+		s := value.(string)
+		return s, nil
+	case integerVariation:
+		f64 := value.(float64)
+		i64 := int64(f64)
+		if f64 == float64(i64) {
+			return i64, nil
+		} else {
+			return nil, fmt.Errorf("failed to convert number to integer")
+		}
+	case numericVariation:
+		number := value.(float64)
+		return number, nil
+	case booleanVariation:
+		v := value.(bool)
+		return v, nil
+	case jsonVariation:
+		v := value.(string)
+		var result interface{}
+		err := json.Unmarshal([]byte(v), &result)
+		if err != nil {
+			return nil, err
+		}
+		return result, nil
+	default:
+		panic("unexpected variationType")
```

Comment:
this could leak into the user's program - would returning an error be sufficient?